himportscan/src/HImportScan/ImportScanner.hs

```diff
@@ -69,26 +65,55 @@ instance Aeson.ToJSON ModuleImport where
 scanImportsFromFile :: FilePath -> IO (Maybe ScannedImports)
 scanImportsFromFile filePath = do
   fileExists <- doesFileExist filePath
-  if fileExists
-  then Just . scanImports filePath <$> Text.readFile filePath
-  else pure Nothing
-
-scanImports :: FilePath -> Text -> ScannedImports
-scanImports filePath contents =
-  let preprocessedContents = Text.encodeUtf8 $ preprocessContents contents
-      sbuffer = case preprocessedContents of
-        PS ptr offset len -> StringBuffer ptr len offset
-      loc = mkRealSrcLoc (mkFastString filePath) 1 1
-   in case scanTokenStream filePath $ lexTokenStream sbuffer loc of
-    Left err -> error err
-    Right ScannedData{moduleName, importedModules, usesTH} ->
-      ScannedImports
-        { filePath = Text.pack filePath
-        , moduleName
-        , importedModules
-        , usesTH
-        }
-
+  if not fileExists
+  then pure Nothing
+  else scanImports filePath =<< Text.readFile filePath
+
+-- TODO[GL]: This function is only in IO because
+-- * we use printBagOfErrors to report an error, but we can easily factor that out
+-- * getImports is in IO, which in turn is only in IO to throw an error
+--   Perhaps we could raise an issue at ghc to make a pure variant.
+scanImports :: FilePath -> Text -> IO (Maybe ScannedImports)
+scanImports filePath contents = do
+  -- TODO[GL]: going through String just because StringBuffer doesn't have a Text interface is not the best
+  -- we could potentially skip that with more effort (e.g. go through ByteString, which is very similar to a StringBuffer)
+  let sb = GHC.stringToStringBuffer $ Text.unpack $ preprocessContents contents
```

Comment:
Is this the same as the old code
```Haskell
sb = case preprocessContents contents of
        PS ptr offset len -> StringBuffer ptr len offset
```
?

If so, why go through an intermediate conversion to `String` here?

Reply:
I guess we can instead use `ByteString.readFile` and rewrite the processing functions to use `ByteString`. But we have to then be careful about unicode :thinking:

Reply:
I'd refrain from that while we don't have performance problems.

Reply:
Avoided the string conversion in fd0fb8d